File: src/main/java/me/zeroeightsix/kami/module/modules/utils/BlockData.java
package me.zeroeightsix.kami.module.modules.utils;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.util.Wrapper;
import net.minecraft.block.Block;
import net.minecraft.block.state.IBlockState;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.RayTraceResult;
import net.minecraftforge.fml.common.gameevent.InputEvent;
import org.lwjgl.input.Mouse;

import java.util.Objects;

/**
 * @author 0x2E | PretendingToCode
 *
 * TODO: Fix delay timer because that shit broken
 * TODO: Move this back to proper category
 */
@Module.Info(name = "BlockData", category = Module.Category.EXPERIMENTAL, description = "Right click blocks to display their data")
public class BlockData extends Module {
    private int delay = 0;

    @Override
    public void onUpdate() {
        if (delay > 0) {
            delay--;
        }
    }

    @EventHandler
    public Listener<InputEvent.MouseInputEvent> mouseListener = new Listener<>(event -> {
        if (Mouse.getEventButton() == 1 && delay == 0) {
            if (mc.objectMouseOver.typeOfHit.equals(RayTraceResult.Type.BLOCK)) {
                BlockPos blockpos = mc.objectMouseOver.getBlockPos();
                IBlockState iblockstate = mc.world.getBlockState(blockpos);
                Block block = iblockstate.getBlock();

                if (block.hasTileEntity()) {
                    TileEntity t = mc.world.getTileEntity(blockpos);
                    NBTTagCompound tag = new NBTTagCompound();
                    Objects.requireNonNull(t).writeToNBT(tag);

                    Command.sendChatMessage(getChatName() + "&6&lBlock Tags:\n" + tag + "");
                }
            }
        }
    });
}

File: src/main/kotlin/org/kamiblue/client/module/modules/misc/BlockData.kt
package org.kamiblue.client.module.modules.misc

import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.RayTraceResult
import net.minecraftforge.fml.common.gameevent.InputEvent
import org.kamiblue.client.module.Category
import org.kamiblue.client.module.Module
import org.kamiblue.client.util.TickTimer
import org.kamiblue.client.util.text.MessageSendHelper
import org.kamiblue.event.listener.listener
import org.lwjgl.input.Mouse

internal object BlockData : Module(
    name = "BlockData",
    category = Category.MISC,
    description = "Right click blocks to display their data"
) {
    private val timer = TickTimer()
    private var lastPos = BlockPos.ORIGIN

    init {
        listener<InputEvent.MouseInputEvent> {
            if (Mouse.getEventButton() != 1 || mc.objectMouseOver == null || mc.objectMouseOver.typeOfHit != RayTraceResult.Type.BLOCK) return@listener
            if (timer.tick(5000L) || mc.objectMouseOver.blockPos != lastPos && timer.tick(500L)) {
                val blockPos = mc.objectMouseOver.blockPos
                val blockState = mc.world.getBlockState(blockPos)
                lastPos = blockPos

                if (blockState.block.hasTileEntity(blockState)) {
                    val tileEntity = mc.world.getTileEntity(blockPos) ?: return@listener
                    val tag = NBTTagCompound().apply { tileEntity.writeToNBT(this) }
                    MessageSendHelper.sendChatMessage("""$chatName &6Block Tags:$tag""".trimIndent())
                }
            }
        }
    }
}